File: js-port/walya-js/src/funcs/keyvalue/kv-select-function.js
const { BaseFunction, CallType } = require('../../core/function-base');
const { SimpleKeyValueCollection } = require('../../model/key-value-collection');
const helpers = require('../helpers');

class KvSelectFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'Select';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 2;
  }

  evaluate(provider, parameters) {
    const error = helpers.expectParamCount(this.symbol, parameters, this.maxParameters);
    if (error) {
      return error;
    }

    const sourceTyped = helpers.ensureTyped(parameters.getParameter(provider, 0));
    if (helpers.typeOf(sourceTyped) !== helpers.FSDataType.KeyValueCollection) {
      return helpers.makeError(
        helpers.FsError.ERROR_TYPE_MISMATCH,
        `${this.symbol} function: The first parameter should be ${this.parName(0)}`
      );
    }

    const selectionTyped = helpers.ensureTyped(parameters.getParameter(provider, 1));
    if (helpers.typeOf(selectionTyped) !== helpers.FSDataType.KeyValueCollection) {
      return helpers.makeError(
        helpers.FsError.ERROR_TYPE_MISMATCH,
        `${this.symbol} function: The second parameter should be ${this.parName(1)}`
      );
    }

    const source = helpers.valueOf(sourceTyped);
    const selection = helpers.valueOf(selectionTyped);
    const projectedEntries = [];

    for (const [key, value] of selection.getAll()) {
      let resolved = value;
      if (!resolved || helpers.typeOf(resolved) === helpers.FSDataType.Null) {
        resolved = source.get(key);
      }
      projectedEntries.push([key, resolved]);
    }

    return helpers.makeValue(
      helpers.FSDataType.KeyValueCollection,
      new SimpleKeyValueCollection(provider, projectedEntries)
    );
  }

  parName(index) {
    return index === 0 ? 'Source KVC' : index === 1 ? 'Target KVC' : '';
  }
}

module.exports = {
  KvSelectFunction
};

File: FuncScript/Functions/KeyValue/KvSelectFunction.cs
using FuncScript.Core;
using FuncScript.Model;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FuncScript.Functions.KeyValue
{
    internal class KvSelectFunction : IFsFunction
    {
        public int MaxParsCount => 2;

        public CallType CallType => CallType.Prefix;

        public string Symbol => "Select";

        public int Precedence => 0;

        public object Evaluate(IFsDataProvider parent, IParameterList pars)
        {
            if (pars.Count != MaxParsCount)
                throw new Error.TypeMismatchError($"{Symbol} function: Invalid parameter count. Expected {MaxParsCount}, but got {pars.Count}");

            var par0 = pars.GetParameter(parent, 0);
            var par1 = pars.GetParameter(parent, 1);

            if (!(par0 is KeyValueCollection))
                throw new Error.TypeMismatchError($"{Symbol} function: The first parameter should be {ParName(0)}");

            if (!(par1 is KeyValueCollection))
                throw new Error.TypeMismatchError($"{Symbol} function: The second parameter should be {ParName(1)}");

            var first = (KeyValueCollection)par0;
            var second = ((KeyValueCollection)par1).GetAll();

            for (int i = 0; i < second.Count; i++)
            {
                if (second[i].Value == null)
                {
                    var key = second[i].Key.ToLower();
                    var value = first.Get(key);
                    second[i] = new KeyValuePair<string, object>(second[i].Key, value);
                }
            }

            return new SimpleKeyValueCollection(parent,second.ToArray());
        }

        public string ParName(int index)
        {
            switch (index)
            {
                case 0: return "Source KVC";
                case 1: return "Target KVC";
            }
            return null;
        }
    }
}
